File: algorithms/lab07/python/03-ce/sol1_tarjan.py
#!/usr/bin/env python3
# SPDX-License-Identifier: BSD-3-Clause

import sys


sys.setrecursionlimit(1_000_000)


class Task:
    def __init__(self):
        # n = numar de noduri, m = numar de muchii/arce
        self.n = 0
        self.m = 0

        # adj[node] = lista de adiacenta a nodului node
        # exemplu: daca adj[node] = {..., neigh, ...} => exista muchia (node, neigh)
        self.adj = []

        # parent[node] = parent of node in the DFS traversal
        self.parent = []

        # found[node] = the timestamp when node was found (when started to visit its subtree)
        # Note: The global timestamp is incremented everytime a node is found.
        self.found = []

        # the minimum accessible timestamp that node can see/access
        # low_link[node] =  min { found[x] | x is node OR x in ancestors(node) OR x in descendants(node) }
        self.low_link = []

    def solve(self) -> None:
        self.read_input()
        self.write_output(self.tarjan_ce())

    def read_input(self) -> None:
        with open("in", "r", encoding="utf-8") as fin:
            self.n, self.m = map(int, fin.readline().split())
            self.adj = [[] for _ in range(self.n + 1)]
            for _ in range(self.m):
                x, y = map(int, fin.readline().split())
                self.adj[x].append(y)
                self.adj[y].append(x)

    def tarjan_ce(self) -> list[tuple[int, int]]:
        # STEP 1: initialize results
        self.parent = [0] * (self.n + 1)
        self.found = [-1] * (self.n + 1)
        self.low_link = [-1] * (self.n + 1)

        # STEP 2: visit all nodes
        all_ces: list[tuple[int, int]] = []
        timestamp = 0
        for node in range(1, self.n + 1):
            if self.parent[node] == 0:
                # node not visited
                # convention: the parent of the root is actually the root
                self.parent[node] = node

                # STEP 3: start a new DFS traversal this subtree
                timestamp = self.dfs(node, timestamp, all_ces)

        return all_ces

    def dfs(self, node: int, timestamp: int, all_ces: list[tuple[int, int]]) -> int:
        # STEP 1: a new node is visited - increment the timestamp
        timestamp += 1
        self.found[node] = timestamp  # the timestamp when node was found
        self.low_link[node] = self.found[node]  # node only knows its timestamp

        # STEP 2: visit each neighbour
        for neigh in self.adj[node]:
            # STEP 3: check if neigh is already visited
            if self.parent[neigh] != 0:
                # STEP 3.1: update low_link[node] with information gained through neigh
                # note: because it's an undirected graph, we should ignore the edge to the parent
                # (the found value of the parent is always less than found value of node)
                if neigh != self.parent[node]:
                    self.low_link[node] = min(self.low_link[node], self.found[neigh])
                continue

            # STEP 4: save parent
            self.parent[neigh] = node

            # STEP 5: recursively visit the child subtree
            timestamp = self.dfs(neigh, timestamp, all_ces)

            # STEP 6: update low_link[node] with information gained through neigh
            self.low_link[node] = min(self.low_link[node], self.low_link[neigh])

            # STEP 7: (node, neigh) is a CE if low_link[neigh] > found[node] where neigh in adj[node]
            if self.low_link[neigh] > self.found[node]:
                all_ces.append((node, neigh))

        return timestamp

    def write_output(self, all_ces: list[tuple[int, int]]) -> None:
        with open("out", "w", encoding="utf-8") as fout:
            fout.write(f"{len(all_ces)}\n")
            for x, y in all_ces:
                fout.write(f"{x} {y}\n")


def main() -> None:
    task = Task()
    task.solve()


if __name__ == "__main__":
    main()

File: algorithms/lab07/rust/03-ce/sol1_tarjan.rs
// SPDX-License-Identifier: BSD-3-Clause

use std::cmp::min;
use std::fs;

#[derive(Default)]
struct Task {
	// n = numar de noduri, m = numar de muchii/arce
	n: usize,

	// adj[node] = lista de adiacenta a nodului node
	// exemplu: daca adj[node] = {..., neigh, ...} => exista muchia (node, neigh)
	adj: Vec<Vec<usize>>,

	// parent[node] = parintele nodului node in parcurgerea DFS
	parent: Vec<usize>,

	// found[node] = momentul cand nodul a fost descoperit (cand a inceput vizitarea subarborelui sau)
	// Observatie: timestamp-ul global creste de fiecare data cand este descoperit un nod.
	found: Vec<i32>,

	// cel mai mic timestamp accesibil pe care il poate vedea/accesa nodul
	// low_link[node] = min { found[x] | x este node SAU x este in stramosii(node) SAU x este in descendentii(node) };
	low_link: Vec<i32>,
}

impl Task {
	fn solve(&mut self) {
		self.read_input();
		let all_ces = self.tarjan_ce();
		self.write_output(&all_ces);
	}

	fn read_input(&mut self) {
		let data = fs::read_to_string("in").expect("failed to read input");
		let mut it = data.split_whitespace();

		self.n = it.next().unwrap().parse::<usize>().unwrap();
		let m = it.next().unwrap().parse::<usize>().unwrap();
		self.adj = vec![Vec::new(); self.n + 1];

		for _ in 0..m {
			let x = it.next().unwrap().parse::<usize>().unwrap();
			let y = it.next().unwrap().parse::<usize>().unwrap();
			self.adj[x].push(y);
			self.adj[y].push(x);
		}
	}

	fn tarjan_ce(&mut self) -> Vec<(usize, usize)> {
		// PASUL 1: initializez rezultatele
		self.parent = vec![0; self.n + 1];
		self.found = vec![-1; self.n + 1];
		self.low_link = vec![-1; self.n + 1];

		let mut all_ces = Vec::new();
		let mut timestamp = 0;

		// PASUL 2: vizitez toate nodurile
		for node in 1..=self.n {
			if self.parent[node] == 0 {
				// nod nevizitat
				// conventie: parintele radacinii este chiar radacina
				self.parent[node] = node;

				// PASUL 3: pornesc o noua parcurgere DFS pe acest subarbore
				self.dfs(node, &mut timestamp, &mut all_ces);
			}
		}

		all_ces
	}

	fn dfs(&mut self, node: usize, timestamp: &mut i32, all_ces: &mut Vec<(usize, usize)>) {
		// PASUL 1: un nod nou este vizitat - incrementez timestamp-ul
		*timestamp += 1;
		self.found[node] = *timestamp; // timestamp-ul la care nodul a fost descoperit
		self.low_link[node] = self.found[node]; // nodul cunoaste doar propriul timestamp

		// PASUL 2: vizitez fiecare vecin
		let neighbors = self.adj[node].clone();
		for neigh in neighbors {
			// PASUL 3: verific daca neigh este deja vizitat
			if self.parent[neigh] != 0 {
				// PASUL 3.1: actualizez low_link[node] cu informatiile obtinute prin neigh
				// observatie: pentru ca graful este neorientat, trebuie ignorata muchia catre parinte
				// (valoarea found a parintelui este mereu mai mica decat valoarea found a lui node)
				if neigh != self.parent[node] {
					self.low_link[node] = min(self.low_link[node], self.found[neigh]);
				}
				continue;
			}

			// PASUL 4: salvez parintele
			self.parent[neigh] = node;

			// PASUL 5: vizitez recursiv subarborele copilului
			self.dfs(neigh, timestamp, all_ces);

			// PASUL 6: actualizez low_link[node] cu informatiile obtinute prin neigh
			self.low_link[node] = min(self.low_link[node], self.low_link[neigh]);

			// PASUL 7: (node, neigh) este un CE daca low_link[neigh] > found[node], unde neigh apartine lui adj[node]
			if self.low_link[neigh] > self.found[node] {
				all_ces.push((node, neigh));
			}
		}
	}

	fn write_output(&self, all_ces: &[(usize, usize)]) {
		let mut out = String::new();
		out.push_str(&format!("{}\n", all_ces.len()));
		for &(x, y) in all_ces {
			out.push_str(&format!("{} {}\n", x, y));
		}
		fs::write("out", out).expect("failed to write output");
	}
}

fn main() {
	let mut task = Box::new(Task::default());
	task.solve();
}
